text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,236 @@
+
+This RFC solves the problem by enabling a parachain to provide a core index commitment as part of it's PVF execution output and in the associated candidate receipt data structure. 
+
+Once this RFC is implemented the validity of a parachain block depends on the core it is being executed on.
+
+## Stakeholders
+
+- Polkadot core developers.
+- Cumulus node developers.
+- Tooling, block explorer developers.
+
+This approach and alternatives have been considered and discussed in [this issue](https://github.com/polkadot-fellows/RFCs/issues/92).
+
+## Explanation
+
+The approach proposed below was chosen primarly because it minimizes the number of breaking changes, the complexity and takes less implementation and testing time. The proposal is to change the existing primitives while keeping binary compatibility with the older versions. We repurpose unused fields to introduce core index and a session index information in the `CandidateDescriptor` and extend the UMP usage to output core index information. 
+
+### Reclaiming unused space in the descriptor
+The `CandidateDescriptor` currently includes `collator` and `signature` fields. The collator includes a signature on the following descriptor fields: parachain id, relay parent, validation data hash, validation code hash and the PoV hash.
+
+However, in practice, having a collator signature in the receipt on the relay chain does not provide any benefits as there is no mechanism to punish or reward collators that have provided bad parachain blocks.
+
+This proposal aims to remove the collator signature and all the logic that checks the collator signatures of candidate receipts. We use the first 6 reclaimed bytes to represent the core and session index, and fill the rest with zeroes. So, there is no change in the layout and length of the receipt. The new primitive is binary compatible with the old one.
+
+### Backwards compatibility
+
+There are two flavors of candidate receipts which are used in network protocols, runtime and node implementation:
+- `CommittedCandidateReceipt` which includes the `CanidateDescriptor` and the `CandidateCommitments` 
+- `CandidateReceipt` which includes the `CanidateDescriptor` and just a hash of the commitments
+
+We want to support both the old and new versions in the runtime and node. The implementation must be able to detect the version of a given candidate receipt.
+
+`CandidateDescriptor` is at version 2 if:
+- the reserved fields are zeroed
+- the session index matches the session index of the relay parent
+- the UMP queue contains a core index commitment and it the one in the descriptor.
```

Comment:
For simplicity we should have it mandatory. I believe it is possible to implement this without requiring any changes from the parachain devs on top of updating the parachain system crate.

Reply:
Why is it simpler? Having a check, if present then must be equal, if not present, don't check .... sounds easy enough.

Reply:
> Reasoning: There is no benefit in enforcing the existence of the commitment, 

I agree from relay chain perspective it doesn't really matter for single core parachains.

>but there are downsides: E.g. A parachain where developers forgot about the commitment would be completely stalled, while if we are not enforcing the existence, it would always be able to operate on at least one core.

Making the core index commitment optional adds an additional knob for elastic scaling parachains, as they will have to enable the commitments in the runtime and collator node logic to parse the `SelectCore` commitment into a core index to put in the descriptor instead of being the default. There is nothing to forget as a parachain dev because all of this logic should be handled by the parachain system pallet in the runtime and `collation-generation` subsystem on the node side. Just upgrading polkadot-sdk and omninode should do the trick.